### PR TITLE
Make OT binding support explicit timestamps.

### LINKED ISSUE #1126: OpenTracing binding ignores start_time argument
OpenTracing specs allow passing explicit timestamps when starting new spans, but in the OT binding  the APM agent this argument is being ignored:
https://github.com/elastic/apm-agent-python/blob/9b360e0c466974df974f5da4f0e34af0735a9942/elasticapm/contrib/opentracing/tracer.py#L79

This issue addressed in this PR already: #824

### PR DESCRIPTION
**Pull request by AndreiPashkin**
## What does this pull request do?

OpenTracing API allows to explicitly set start and end timestamps of spans but the implementation in this project does not allow that. That patch fixes it.

Closes #1126

### CODE REVIEW

**Comment by basepi** [comment]
jenkins run the tests please

**Comment by basepi** [comment]
Also, @AndreiPashkin, please sign the [CLA](https://www.elastic.co/contributor-agreement) when you get a chance.

**Comment by AndreiPashkin** [comment]
> Also, @AndreiPashkin, please sign the [CLA](https://www.elastic.co/contributor-agreement) when you get a chance.

Hmm, but I've signed it, here is the PDF:
https://www.dropbox.com/s/oufsg1mc17hlqvy/Please_DocuSign_Individual_Contributor_Licen%20%281%29.pdf?dl=0

Why doesn't the check turn green?

I'm also puzzled - how to I can I run tests so that I can use PDB for debugging locally? Could you please advise? I'd like to add one test case.

**Comment by basepi** [comment]
Looks like CLA went green, maybe it was delayed.

> I'm also puzzled - how can I run tests so that I can use PDB for debugging locally?

I don't tend to use PDB, but running the tests should be as simple as installing the [requirements file](https://github.com/elastic/apm-agent-python/tree/master/tests/requirements) that is most relevant (at minimum, [reqs-base.txt](https://github.com/elastic/apm-agent-python/blob/master/tests/requirements/reqs-base.txt)), and then run `py.test` to run all the tests. I'm guessing PDB can attach to pytest runs without issue -- most of my debugging is either via logging or via the VSCode debugger.

**Comment by AndreiPashkin** [comment]
Let me actually write a test.

**Comment by AndreiPashkin** [comment]
@basepi, @beniwohli, I'm observe some strange things happening with these changes in place.

Here is my test-script:
```
import logging
import sys
import time


from elasticapm import Client
from elasticapm.contrib.opentracing import Tracer


client = Client({
    'SERVICE_NAME': 'test-with_changes',
    'SERVER_URL': 'http://localhost:8888'
})
tracer = Tracer(client_instance=client)

logging.basicConfig(stream=sys.stderr, level=logging.DEBUG)

START = time.time()

span_1 = tracer.start_span('parent', start_time=START)
span_2 = tracer.start_span('child', child_of=span_1, start_time=START + 2)
span_2.finish(finish_time=START + 10)
span_1.finish(finish_time=START + 15)
```

When I install latest release with `pip install -I elastic-apm` and execute it - I see such a picture:
image
Both parent and child are visible but they have incorrect times recorded (and making it recorded correctly is the purpose of this PR).

When I install my changes (`pip install -I https://github.com/AndreiPashkin/apm-agent-python/archive/explicit-timestamps-opentracing.zip`) and re-running my script (changing the service name to better observe the differences) - I see the following:
image
And the sent payload:
```
{"metadata": {"service": {"name": "test-with_changes", "environment": null, "version": null, "agent": {"name": "python", "version": "5.6.0"}, "language": {"name": "python", "version": "3.7.7"}, "runtime": {"name": "CPython", "version": "3.7.7"}}, "process": {"pid": 31801, "ppid": 2071, "argv": ["test_elastic.py"], "title": null}, "system": {"hostname": "andrei-yoga514", "architecture": "x86_64", "platform": "linux"}}}
{"span": {"id": "a1a2d1d77c17595e", "transaction_id": "0af7966685c1e27a", "trace_id": "7adccfb5c60bad87b80e290334e6ee53", "parent_id": "0af7966685c1e27a", "name": "child", "type": "code", "subtype": "custom", "action": null, "sync": true, "timestamp": 3179173112604596, "duration": 8000.0, "stacktrace": [{"abs_path": "test_elastic.py", "filename": "test_elastic.py", "module": "__main__", "function": "<module>", "lineno": 21, "library_frame": false}]}}
{"transaction": {"id": "0af7966685c1e27a", "trace_id": "7adccfb5c60bad87b80e290334e6ee53", "name": "parent", "type": "custom", "duration": 1589530226905.7947, "result": "", "timestamp": 1589642898698801, "sampled": true, "span_count": {"started": 1, "dropped": 0}, "context": {"tags": {}}}}
```

1. Only parent is seen.
2. Timings have gigantic values.

I've also captured payload that were sent in both cases.

Without changes:
```
{"metadata": {"service": {"name": "test-without_changes", "environment": null, "version": null, "agent": {"name": "python", "version": "5.6.0"}, "language": {"name": "python", "version": "3.7.7"}, "runtime": {"name": "CPython", "version": "3.7.7"}}, "process": {"pid": 30721, "ppid": 2071, "argv": ["test_elastic.py"], "title": null}, "system": {"hostname": "andrei-yoga514", "architecture": "x86_64", "platform": "linux"}}}
{"span": {"id": "da6e65e8787355b7", "transaction_id": "f5c2f6d6740a0cd4", "trace_id": "4366c9a4a3cbc6c54abbf84e10d4d8a5", "parent_id": "f5c2f6d6740a0cd4", "name": "child", "type": "code", "subtype": "custom", "action": null, "sync": true, "timestamp": 1589642748818084, "duration": 0.07035498856566846}}
{"transaction": {"id": "f5c2f6d6740a0cd4", "trace_id": "4366c9a4a3cbc6c54abbf84e10d4d8a5", "name": "parent", "type": "custom", "duration": 0.24842699349392205, "result": "", "timestamp": 1589642748818004, "sampled": true, "span_count": {"started": 1, "dropped": 0}, "context": {"tags": {}}}}
```
With changes:
```
{"metadata": {"service": {"name": "test-with_changes", "environment": null, "version": null, "agent": {"name": "python", "version": "5.6.0"}, "language": {"name": "python", "version": "3.7.7"}, "runtime": {"name": "CPython", "version": "3.7.7"}}, "process": {"pid": 31801, "ppid": 2071, "argv": ["test_elastic.py"], "title": null}, "system": {"hostname": "andrei-yoga514", "architecture": "x86_64", "platform": "linux"}}}
{"span": {"id": "a1a2d1d77c17595e", "transaction_id": "0af7966685c1e27a", "trace_id": "7adccfb5c60bad87b80e290334e6ee53", "parent_id": "0af7966685c1e27a", "name": "child", "type": "code", "subtype": "custom", "action": null, "sync": true, "timestamp": 3179173112604596, "duration": 8000.0, "stacktrace": [{"abs_path": "test_elastic.py", "filename": "test_elastic.py", "module": "__main__", "function": "<module>", "lineno": 21, "library_frame": false}]}}
{"transaction": {"id": "0af7966685c1e27a", "trace_id": "7adccfb5c60bad87b80e290334e6ee53", "name": "parent", "type": "custom", "duration": 1589530226905.7947, "result": "", "timestamp": 1589642898698801, "sampled": true, "span_count": {"started": 1, "dropped": 0}, "context": {"tags": {}}}}
```

**Comment by AndreiPashkin** [comment]
@beniwohli, can you also please explain this code:
http://github.com/elastic/apm-agent-python/blob/master/elasticapm/traces.py#L140-L140

Why if `start` provided both `self.timestamp` and `self.start_time` are assigned to it but if `start` is not provided `self.timestamp` is assigned epoch timestamp and `self.start_time` is assigned result of `time.perf_time()` which is usually equals to `time.monotonic()`?

File: elasticapm/utils/time.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        return measurement
+
+
+    def __check_measurements_cache(self, tolerance):
+        """Check if there are cached clock measurements."""
+        if self.__clocks_measurement_cache is None:
+            return
+        if self.__measurements_ttl is None or self.__measurements_ttl == 0:
+            return
+        if ((self.__clocks_measurement_cache[0] + self.__measurements_ttl) >
```
[the remark this answers is not in the document]
Comment (AndreiPashkin) [reply elasticapm/utils/time.py:81]: @basepi, I'm not sure, how should I re-format this line? Without breaks it is going to become too long.

[the remark this answers is not in the document]
Comment (basepi) [reply elasticapm/utils/time.py:80]: I'm a little surprised that `black` isn't reformatting this for you...

Our enforced line length limit is actually 120 columns, so I think you can just do this on one line. I try to keep my code down to around 90 columns, but if it hurts readability I think it's not worth chasing.

File: elasticapm/utils/time.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            )
+        return (
+            clocks_measurements.destination +
+            (timestamp - clocks_measurements.source)
+        )
+
+
+SYSTEM_TO_MONOTONIC = TimestampConversionUtil(
+    time.time,
+    time.monotonic,
```
Comment (beniwohli) [inline elasticapm/utils/time.py:114]: `time.monotonic` is not available on Python 2.7. To ensure Python 2.7 compatibility, `timeit.default_timer` is the best option. On Python 3, it defaults to `time.perf_counter`, on Python 2 to the best platform dependent time source (non-monotonic)

**Comment by AndreiPashkin** [comment]
@beniwohli
>  it's not bulletproof as far as I can tell (e.g. the system clock could jump right after you measured the delta, but before applying it to the non-monotonic timestamp).
But if system clocks is jumping than the agent gonna send skewed data in any case - so there is no bullet-proofness against that in current `master` anyway and my PR does not change that. `TimestampConversionUtil` simply converts timestamps between clocks without adding or removing anything to protection against clock jumps.

> As such, I'm a bit wary to introduce this into a major code path of the agent. However, I'd be open to it if you moved the usage of TimestampConversionUtil to the opentracing bridge.
Should I add another private argument for system-clock timestamp and pass it from the OT bridge along with the monotonic timestamp?

**Comment by AndreiPashkin** [comment]
@beniwohli
Could you please reply on my prev. comment?

**Comment by AndreiPashkin** [comment]
@basepi, @beniwohli, could you please look into it?

**Comment by basepi** [comment]
@beniwohli As far as I can tell, the new `TimestampConversionUtil` is only used within the OT bridge, and even then only if a start time is defined, so I don't think this is actually going to introduce that additional complexity across core transaction/span duration measurements. 

For my sanity, @AndreiPashkin, can you give me the tl;dr of why that class is necessary? Are we just trying to avoid the unlikely, but possible chance that we could measure the difference between the clocks at the point of a daylight saving change, and thus convert with an hour mistake in the duration? Are there other examples of jumps in non-monotonic clocks that we're trying to detect/prevent?

**Comment by basepi** [comment]
Closing this due to lack of reply.

**Pull request closed without merging**